enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,728 @@
+
+#### Kubelet
+
+In OpenShift, `machine-config-operator` manages `kubelet` configuration.
+Based on the [Infrastructure value](https://github.com/openshift/api/blob/master/config/v1/types_infrastructure.go) MCO currently sets the correct `cloud-config` value.
+To support external providers, MCO should not set the `cloud-config` value anymore.
+
+Additionally, it must set the value of the `cloud-provider` [flag](https://github.com/openshift/machine-config-operator/blob/2a51e49b0835bbd3ac60baa685299e86777b064f/templates/worker/01-worker-kubelet/_base/units/kubelet.service.yaml#L32) to `external`.
+This informs `kubelet` that an out-of-tree provider will finish the `Node` initialization process.
+
+While `kubelet` itself does not directly contact or interact with the `CCM`, formation of a functional `Node` in Kubernetes does rely on the `CCM`.
+With in-tree providers, `kubelet` looks up Node IP Addresses, Instance IDs and other metadata directly from the cloud provider and places this information onto the `Node` object.
+
+When using out-of-tree, the `kubelet` taints the `Node` with `node.cloudprovider.kubernetes.io/uninitialized: NoSchedule` and does not look up any information.
+It is then the responsibility of the `CCM` to look up the Node IP Addresses, Instance IDs and other metadata and labels, place the information onto the `Node` object and then remove the taint from the `Node`.
+
+The information previously set by `kubelet`, now `CCM`, is required to make a `Node` schedulable.
+Without this information, the `scheduler` cannot schedule `Pods` that have any sort of affinity that relies on `Node` metadata (eg failure domain).
+
+To ensure that cluster disaster recovery procedures can still operate smoothly, we will ensure that core control plane components and their operators tolerate the uninitialized taint, to prevent `CCM` blocking new control plane hosts being added if `CCM` is non-functional.
+This will include, but is not limited to: Kube Controller Manager, Etcd, Kube API Server, Networking, Cluster Machine Approver.
+
+#### Windows Machine Config Operator
+
+##### Kubelet Changes
+
+To generate the configuration for Windows nodes, `WMCO` reads the output of MCO's rendered configuration as a basis for its own config for Kubelet on the Windows node.
+As we are making changes to the output of the Kubelet service in `MCO` (namely change the value of the `cloud-provider` flag), we will need to verify that `WMCO` reads this flag and copies its value to the Kubelet Windows service.
+
+##### Node Initialization
+
+On most platforms, `Node` initialization is handled centrally by the `CCM`, specifically the Cloud Node Manager (`CNM`) running within it.
+However, On certain platforms (e.g. Azure), the `CNM` must be run on the `Node` itself, typically via a `DaemonSet`.
+Since Red Hat cannot supply or support Windows container images, we cannot run a `DaemonSet` for the `CNM` targeted at Windows Nodes as we would do on Linux Nodes.
```

Comment:
If WMCO is managing CNM, its worth noting that WMCO is managed by OLM and not CVO, so Windows nodes will updated independently of the rest of the cluster.
This means there will be version skew for some period of time, especially if WMCO upgrades are set to be manually approved by the cluster administrator.
I'm not sure if this would cause an issue, but worth keeping in mind.

Reply:
Added a detail there.